### PR TITLE
Feat: Tenant Flow

### LINKED ISSUE #70: Tenant Flow -  Implement user registration flow
# [Frontend] Tenant Flow - Implement user registration flow

Implement the tenant registration process using Firebase Auth and Hasura to store role information.

## Problem
We need to:
- Register new users through Firebase
- Assign the 'tenant' role in our database
- Handle email verification
- Create proper onboarding flow

### GraphQL Operations

```graphql
# After Firebase creates the user, assign tenant role
mutation AssignTenantRole($userId: String!) {
  insert_user_roles_one(object: {
    user_id: $userId,
    role_id: 2  # Assuming 2 is tenant role id
  }) {
    user_id
    role {
      name
    }
  }
}

# Create initial user record
mutation CreateUserRecord($user: users_insert_input!) {
  insert_users_one(object: $user) {
    id
    email
    last_seen
  }
}
```

### React Implementation

```tsx
// src/features/auth/TenantRegistration.tsx
import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { 
  getAuth, 
  createUserWithEmailAndPassword, 
  sendEmailVerification 
} from 'firebase/auth';
import { useMutation } from '@apollo/client';
import { ASSIGN_TENANT_ROLE, CREATE_USER_RECORD } from './graphql';

export const TenantRegistration: React.FC = () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState<string | null>(null);
  const navigate = useNavigate();
  
  const [assignRole] = useMutation(ASSIGN_TENANT_ROLE);
  const [createUser] = useMutation(CREATE_USER_RECORD);
  
  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    setError(null);
    
    try {
      // 1. Create Firebase user
      const auth = getAuth();
      const { user } = await createUserWithEmailAndPassword(
        auth,
        email,
        password
      );
      
      // 2. Send email verification
      await sendEmailVerification(user);
      
      // 3. Create user record in our database
      await createUser({
        variables: {
          user: {
            id: user.uid,
            email: user.email,
          }
        }
      });
      
      // 4. Assign tenant role
      await assignRole({
        variables: {
          userId: user.uid
        }
      });
      
      // 5. Redirect to verification page
      navigate('/verify-email');
      
    } catch (error) {
      setError(getErrorMessage(error));
    }
  };
  
  return (
    <div className="max-w-md mx-auto p-6">
      <h1 className="text-2xl font-bold mb-6">
        Register as Tenant
      </h1>
      
      {error && (
        <div className="bg-red-50 text-red-500 p-3 rounded mb-4">
          {error}
        </div>
      )}
      
      <form onSubmit={handleSubmit} className="space-y-4">
        <div>
          <label className="block mb-1">Email</label>
          <input
            type="email"
            value={email}
            onChange={(e) => setEmail(e.target.value)}
            className="w-full p-2 border rounded"
            required
          />
        </div>
        
        <div>
          <label className="block mb-1">Password</label>
          <input
            type="password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
            className="w-full p-2 border rounded"
            minLength={6}
            required
          />
        </div>
        
        <button
          type="submit"
          className="w-full bg-blue-500 text-white p-2 rounded"
        >
          Register
        </button>
      </form>
    </div>
  );
};

// Helper function to parse Firebase errors
const getErrorMessage = (error: any): string => {
  switch (error.code) {
    case 'auth/email-already-in-use':
      return 'This email is already registered';
    case 'auth/invalid-email':
      return 'Invalid email address';
    case 'auth/operation-not-allowed':
      return 'Email/password accounts are not enabled';
    case 'auth/weak-password':
      return 'Password should be at least 6 characters';
    default:
      return 'An error occurred during registration';
  }
};
```

### Email Verification Component

```tsx
// src/features/auth/EmailVerification.tsx
import { useEffect, useState } from 'react';
import { getAuth, onAuthStateChanged } from 'firebase/auth';
import { useNavigate } from 'react-router-dom';

export const EmailVerification: React.FC = () => {
  const [isVerified, setIsVerified] = useState(false);
  const navigate = useNavigate();
  
  useEffect(() => {
    const auth = getAuth();
    const unsubscribe = onAuthStateChanged(auth, async (user) => {
      if (user) {
        // Reload user to get latest email verification status
        await user.reload();
        if (user.emailVerified) {
          setIsVerified(true);
          // Wait a moment then redirect
          setTimeout(() => {
            navigate('/dashboard');
          }, 2000);
        }
      }
    });
    
    return () => unsubscribe();
  }, [navigate]);
  
  return (
    <div className="max-w-md mx-auto p-6 text-center">
      <h1 className="text-2xl font-bold mb-4">
        Verify Your Email
      </h1>
      
      {isVerified ? (
        <div className="text-green-500">
          Email verified! Redirecting...
        </div>
      ) : (
        <div className="space-y-4">
          <p>
            Please check your email and click the verification link.
          </p>
          <p className="text-sm text-gray-500">
            Don't see the email? Check your spam folder.
          </p>
        </div>
      )}
    </div>
  );
};
```

## Success Flow
1. User enters email and password
2. Firebase creates authentication record
3. Verification email is sent
4. User record is created in our database
5. Tenant role is assigned
6. User is redirected to verification page
7. After verification, user accesses dashboard

## Error Handling
- Email already registered
- Invalid email format
- Weak password
- Network issues
- Database errors
- Role assignment failures

## Testing Scenarios
1. Successful registration flow
2. Email verification flow
3. Invalid email/password combinations
4. Network error handling
5. Database error handling
6. Navigation flows
7. Protected route access

## Prerequisites
- [x] Firebase Auth configured (#9)
- [x] Hasura JWT setup completed
- [x] User roles table created
- [ ] Frontend routes configured
- [ ] Error boundaries implemented

## Success Criteria
- [ ] User can register with email/password
- [ ] User record is created in database
- [ ] Tenant role is assigned correctly
- [ ] Error states are handled properly
- [ ] Loading states are implemented
- [ ] Proper navigation flow works

## Additional Notes
- Consider adding password strength indicator
- Implement rate limiting for registration attempts
- Add terms of service acceptance
- Consider adding social login options
- Add proper loading states
- Implement proper form validation

## Related Issues

-[ Firebase Auth Integration](https://github.com/safetrustcr/Backend/issues/22)

### PR DESCRIPTION
**Pull request by lauchaves**
# Pull Request for SafeTrust - Close Issue #70 

❗ **Pull Request Information**

Implement the tenant registration process using Firebase Auth and Hasura to store role information.

## 🌀 Summary of Changes

- **UI Enhancements:**
  - Improve login UI and mobile views.
  - Created the tenant register view matching the current styles
  - Refactor auth buttons so it's a reusable `AuthButtons` component with Google and Wallet login options.
  - Dynamically displayed login/register button text based on the action.

- **Terms of Service:**
  - Added a checkbox to accept Terms of Service before submitting the registration/login form.
  - Included a link to the full Terms of Service ( page not implemented yet), form validate that its checked before submitting it.

- **Password Validation:**
  - Implemented password strength validation

- **Field Validation:**
  - Improved form validation for email, password

- **Error Handling:**
  - Added error handling

- **Firebase Authentication for Social Auths:**
  - Initialized Google sign-in with `signInWithPopup()` using Firebase's `GoogleAuthProvider`.
  - Added wallet authentication support to MetaMask using `ethers.js`.
  
- **Update Readme**
  - Updated the readme with the proper steps to set up the repo

### Evidence Before Solution

- There was no tenant flow. 
- Previous login mobile view:
image

### Evidence After Solution

- **Video**: [Link to Loom video](https://www.loom.com/share/4dce713ab7514458af031a4573ba129f?sid=c5696d5e-878c-4548-b029-ee7faf851092)

-**Screenshots**
Wallet connection for auth: 
image
image

## 📂 Related Issue

This pull request will **close #70 ** upon merging.

---

## Author comments:

- This is ready to be merged at any time. I have left TODO comments regarding uncommenting some functions once the Firebase/DB setup is ready. This can be done in a separate issue if needed, to prevent this PR from remaining open for too long.
- Matched the current app styles.
- I updated some styles for the login
- I updated the readme since it was not quite the steps to set up the repo

🎉 Thank you for reviewing this PR! 🎉

### CODE REVIEW

File: src/components/auth/tenant/TenantRegistration.tsx [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import PasswordStrengthIndicator from './PasswordStrengthIndicator';
+
+export const TenantRegistration: React.FC = () => {
+  const [email, setEmail] = useState('');
+  const [password, setPassword] = useState('');
+  const [error, setError] = useState<string | null>(null);
+  const router = useRouter();
+  
+  const [assignRole] = useMutation(ASSIGN_TENANT_ROLE);
+  const [createUser] = useMutation(CREATE_USER_RECORD);
```
Comment (lauchaves) [inline src/components/auth/tenant/TenantRegistration.tsx:21]: @rvalenciano hey 👋 would u mind doing an early a code review Im having this issue 
image

I researched about it and seems to be something related to apollo setup but I'm not sure, do u have any idea what could it be?

**Comment by lauchaves** [comment]
@rvalenciano this is ready for review!

**Comment by JosueBrenes** [comment]
Hi @lauchaves please resolve conflicts branch 

image

**Comment by JosueBrenes** [comment]
Hi @lauchaves please resolve conflicts

image

**Comment by JosueBrenes** [comment]
Hiiii @lauchaves there are still conflicts jajaj
 
image

**Comment by lauchaves** [comment]
>  Hiiii @lauchaves there are still conflicts jajaj
> 
>  
> 
> image
> 
> 

@JosueBrenes that just happened haha I'll address them give me a couple of mins

**Comment by lauchaves** [comment]
@JosueBrenes 👀 
image

**Pull request merged**